src/storage-client/src/ssh_tunnels.rs

```diff
@@ -0,0 +1,233 @@
+
+//! SSH tunnel management for the storage layer.
+
+// NOTE(benesch): The synchronization in this module is tricky because SSH
+// tunnels 1) require an async `connect` method that can return errors and 2)
+// once connected, launch a long-running background task whose handle must be
+// managed. The manager would be far simpler if `connect` was neither async nor
+// fallible and instead synchronously returned a handle to the background task.
+// That would require a different means of asynchronously reporting SSH tunnel
+// errors, though, and that's a large project. A worthwhile project, though: at
+// present SSH tunnel errors that occur after the initial connection are
+// reported only to the logs, and not to users.
+
+use std::collections::{btree_map, BTreeMap};
+use std::ops::Deref;
+use std::sync::{Arc, Mutex};
+
+use futures::{future, FutureExt};
+use scopeguard::ScopeGuard;
+use tokio::sync::oneshot;
+use tracing::info;
+
+use mz_repr::GlobalId;
+use mz_secrets::SecretsReader;
+use mz_ssh_util::keys::SshKeyPairSet;
+use mz_ssh_util::tunnel::{SshTunnelConfig, SshTunnelHandle};
+
+use crate::types::connections::SshTunnel;
+
+/// Thread-safe manager of SSH tunnel connections.
+#[derive(Debug, Clone, Default)]
+pub struct SshTunnelManager {
+    tunnels: Arc<Mutex<BTreeMap<SshTunnelKey, SshTunnelState>>>,
+}
+
+impl SshTunnelManager {
+    /// Establishes an SSH tunnel for the given remote host and port using the
+    /// provided `tunnel` configuration.
+    ///
+    /// If there is an existing SSH tunnel, a handle to that tunnel is returned,
+    /// rather than establishing a new tunnel.
+    ///
+    /// The manager guarantees that there will never be more than one in flight
+    /// connection attempt for the same tunnel, even when this method is called
+    /// concurrently from multiple threads.
+    pub async fn connect(
+        &self,
+        secrets_reader: &dyn SecretsReader,
+        tunnel: &SshTunnel,
+        remote_host: &str,
+        remote_port: u16,
+    ) -> Result<ManagedSshTunnelHandle, anyhow::Error> {
+        // An SSH tunnel connection is uniquely identified by the ID of the
+        // Materialize connection (in the `CREATE CONNECTION` sense) and the
+        // remote address.
+        let key = SshTunnelKey {
+            connection_id: tunnel.connection_id,
+            remote_host: remote_host.to_string(),
+            remote_port,
+        };
+
+        loop {
+            // NOTE: this code is structured awkwardly to convince rustc that
+            // the lock is not held across an await point. rustc's analysis
+            // does not take into account explicit `drop` calls, so we have to
+            // structure such that the lock guard goes out of scope.
+            // See: https://github.com/rust-lang/rust/issues/69663
+            enum Action {
+                Return(ManagedSshTunnelHandle),
+                AwaitConnection(future::Shared<oneshot::Receiver<()>>),
```

Comment:
nit: you might prefer: https://docs.rs/tokio/latest/tokio/sync/watch/index.html here